### PR TITLE
Unify route macros

### PR DESCRIPTION
**Pull request by arniu**
## PR Type

Refactor

## PR Checklist
Check your PR fulfills the following:

- [x] Tests for the changes have been added / updated.
- [x] Documentation comments have been added / updated.
- [ ] A changelog entry has been made for the appropriate packages.
- [x] Format code with the latest stable rustfmt

### CODE REVIEW

**Review by Neopallium — APPROVED** [review]
looks good.

**Review by Neopallium — APPROVED** [review]

File: actix-web-codegen/src/lib.rs [hunk 1]
```diff
@@ -17,10 +17,12 @@
 //! - [options](attr.options.html)
 //! - [trace](attr.trace.html)
 //! - [patch](attr.patch.html)
+//! - [route](attr.route.html)
 //!
 //! ### Attributes:
 //!
-//! - `"path"` - Raw literal string with path for which to register handle. Mandatory.
+//! - `"path"` - *Required*, raw literal string with path for which to register handle
+//! - `method="HTTP_METHOD"` - Secondary HTTP method accepted, uppercased string. "GET", "POST" for example
```
Comment (robjtede) [inline actix-web-codegen/src/lib.rs]: please make it clear this only applies to the route macro

Comment (robjtede) [reply actix-web-codegen/src/lib.rs]: also: uppercased => uppercase or upper-cased

File: actix-web-codegen/src/lib.rs [hunk 2]
```diff
@@ -155,7 +63,52 @@ pub fn patch(args: TokenStream, input: TokenStream) -> TokenStream {
...
+    ) => {
+        $(
+            doc_comment! {
+                concat!("Creates route handler with `", stringify!($upper), "` method guard.")
+                concat!("")
+                concat!("Syntax: `#[", stringify!($method), "(\"path\" [, attributes])]`")
+                concat!("")
+                concat!("## Attributes:")
+                concat!("")
+                concat!("- `\"path\"` - *required* Raw literal string with path for which to register handler")
```
Comment (robjtede) [inline actix-web-codegen/src/lib.rs]: i don't think this required marker is necessary, it's specified in the syntax part

File: actix-web-codegen/src/lib.rs [hunk 3]
```diff
@@ -17,10 +17,12 @@
 //! - [options](attr.options.html)
 //! - [trace](attr.trace.html)
 //! - [patch](attr.patch.html)
+//! - [route](attr.route.html)
 //!
 //! ### Attributes:
```
Comment (robjtede) [inline actix-web-codegen/src/lib.rs]: it's not clear this section should be here any more

File: actix-web-codegen/src/route.rs [hunk 4]
```diff
@@ -20,79 +20,81 @@ impl ToTokens for ResourceType {
...
-    Put,
-    Delete,
-    Head,
-    Connect,
-    Options,
-    Trace,
-    Patch,
-    Multi,
-}
+macro_rules! method_type {
```
Comment (robjtede) [inline actix-web-codegen/src/route.rs:23]: this is a great idea 👍

**Review by robjtede — COMMENTED** [review]
Pretty nice clean up overall. Only major point at the moment is wether the attribute section in the main page docs should exist; possibly should remove it and move the notes section into each individual macro.

File: actix-web-codegen/src/lib.rs (hunk 3 above)
Comment (arniu) [reply actix-web-codegen/src/lib.rs]: Shall I take them all down, or just remove `route`?

Comment (robjtede) [reply actix-web-codegen/src/lib.rs]: yeah i think remove it and move the notes section

File: actix-web-codegen/src/lib.rs (hunk 1 above)
Comment (arniu) [reply actix-web-codegen/src/lib.rs]: Now, other macros can take `method` too.

For example, `put("/form", method="POST")` will be accepted.

Do you think it is a feature, or a bug?

Comment (robjtede) [reply actix-web-codegen/src/lib.rs]: no, that's not desired, multiple methods should only be allowed on route macro

**Comment by robjtede** [comment]
I'd like to make sure the main doc page has some more guidance than this but I'll submit it as the next PR.

**Review by robjtede — APPROVED** [review]

**Pull request merged**